### DISCUSSION TITLE
Range restricted dependency with outdated lockfile causes infinite renovate loop

### DISCUSSION THREAD

**Discussion by Garethp** [renovatebot/renovate#26670] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.133.0

### If you're self-hosting Renovate, select which platform you are using.

gitlab.com

### Was this something which used to work for you, and then stopped?

I have not seen this working

### Describe the problem

Package Manager: `yarn 3.2.3`
Renovate Version: `37.133.0`
Reproduction: https://github.com/Garethp/renovate-loop-bug
Platforms: Gitlab & Github
Self-hosted

When I have a dependency that's caret restricted ([`"is-arrayish": "^0.3.0"`](https://github.com/Garethp/renovate-loop-bug/blob/main/package.json#L5)) that has an outdated version in the `yarn.lock` file ([`version: 0.3.0`](https://github.com/Garethp/renovate-loop-bug/blob/main/yarn.lock#L8)) in the same group as a pinned version that Renovate is trying to update then Renovate gets stuck in a loop where:

1. It creates the branch attempting to update the other dependencies to their latest version
2. On the next run it updates the branch to remove the changes from Step 1 and only updates the lockfile to bump `is-arrayish` and nothing else
3. On a third run it sets the branch back to the state it was in after Step 1
4. Repeat

I encountered this issue running self-hosted on Gitlab Runners, however this issue is easily reproduced on Github by running a self-hosted renovate against it. The Github repo that I've linked is the same repo that I use to run renovate on itself. My methods of reproducing are:

1. Clone the renovate repository
2. `yarn install`
3. `export RENOVATE_TOKEN={githubPersonalAccessToken}`
4. `yarn renovate && yarn renovate`

I attempted older versions of Renovate, and it looked as though this bug went back as far as `35.0.0`, though I didn't check how much further back it went.

### Relevant debug logs

<details><summary>echo "==== RUN 1 ====" >> log.txt && yarn renovate >> log.txt && echo "==== RUN 2 ====" >> log.txt && yarn renovate >> log.txt</summary>

```
==== RUN 1 ====
 INFO: Repository started (repository=Garethp/renovate-loop-bug)
       "renovateVersion": "37.140.0"
 INFO: Dependency extraction complete (repository=Garethp/renovate-loop-bug, baseBranch=main)
       "stats": {
         "managers": {"npm": {"fileCount": 1, "depCount": 2}},
         "total": {"fileCount": 1, "depCount": 2}
       }
 INFO: Branch created (repository=Garethp/renovate-loop-bug, branch=renovate/group-all)
       "commitSha": "283c9db4a3a80d1c77dc5f4e73d1f9889e911358"
 INFO: PR created (repository=Garethp/renovate-loop-bug, branch=renovate/group-all)
       "pr": 8,
       "prTitle": "Update group:all"
 INFO: Repository finished (repository=Garethp/renovate-loop-bug)
       "cloned": true,
       "durationMs": 8420
==== RUN 2 ====
 INFO: Repository started (repository=Garethp/renovate-loop-bug)
       "renovateVersion": "37.140.0"
 INFO: Dependency extraction complete (repository=Garethp/renovate-loop-bug, baseBranch=main)
       "stats": {
         "managers": {"npm": {"fileCount": 1, "depCount": 2}},
         "total": {"fileCount": 1, "depCount": 2}
       }
 INFO: Branch updated (repository=Garethp/renovate-loop-bug, branch=renovate/group-all)
       "commitSha": "ebfdada25d1bebb254f38e63f0d0780b31f7c73f"
 INFO: Repository finished (repository=Garethp/renovate-loop-bug)
       "cloned": true,
       "durationMs": 8148

```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

**Comment by Garethp** [comment]
Is there any further information I could give that could help with this?

**Reply by rarkins** [reply]
This reproduction takes too long to run for it to be feasible for me to debug. Each yarn install locally is taking 2+ minutes and there are multiple branches. Please try to reduce its size, for example by ignoring all deps except the one needed to reproduce the problem.

BTW if branch automerge is a requirement to reproduce then you have forgotten to mention that details and maybe others in your description. If they are not a requirement to reproduce then please remove them from the reproduction. Each config should be required to keep it minimal.

**Reply by Garethp** [reply]
Thanks for the feedback. I've stripped the repo down further with only two dependencies that each have no dependencies of their own. I've also cut the config back as far as I can. The only option I couldn't remove was in `config.js` where I set the `groupName: "group:all"`. This is required because the bug only occurs if both dependencies are in the same group.

**Reply by rarkins** [reply]
I'm getting this error when trying to reproduce it locally:

```
cmd": "/bin/sh -c docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/tmp/renovate/repos/github/renovate-reproductions/26670\":\"/tmp/renovate/repos/github/renovate-reproductions/26670\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -e CI -e YARN_ENABLE_IMMUTABLE_INSTALLS -e YARN_HTTP_TIMEOUT -e YARN_GLOBAL_FOLDER -e YARN_ENABLE_GLOBAL_CACHE -e CONTAINERBASE_CACHE_DIR -w \"/tmp/renovate/repos/github/renovate-reproductions/26670\" ghcr.io/containerbase/sidecar:9.31.4 bash -l -c \"install-tool node v20.11.0 && install-tool yarn 1.22.21 && yarn install --mode=update-lockfile && yarn up -R is-arrayish --mode=update-lockfile\"",
         "stderr": "error Command \"up\" not found.\n",
```

`yarn up` is a Yarn 2+ command, but it seems Renovate is detecting Yarn v1

**Reply by Garethp** [reply]
I'm not sure why it's detecting Yarn 1. I've updated the reproduction to specify the yarn version in the `package.json` and updated it to `Yarn 4.0.2` so that Renovate won't attempt to update yarn

**Reply by rarkins** [reply]
Yes, that fixed the error. Is this branch commit the one you're expecting? https://github.com/renovate-reproductions/26670/commit/3f946fb62f2afcdbb747613b409e8937513c85de

**Reply by rarkins** [reply]
Error is not reproduced on the next run, it behaves as expected:

```
DEBUG: latest commit (repository=renovate-reproductions/26670, branch=renovate/is-arrayish-0.x-lockfile)
       "branchName": "main",
       "latestCommitDate": "2024-01-18T14:14:46+00:00"
DEBUG: manager.getUpdatedPackageFiles() reuseExistingBranch=true (repository=renovate-reproductions/26670, branch=renovate/is-arrayish-0.x-lockfile)
DEBUG: npm.updateLockedDependency: is-arrayish@0.3.0 -> 0.3.2 [yarn.lock] (repository=renovate-reproductions/26670, branch=renovate/is-arrayish-0.x-lockfile)
DEBUG: is-arrayish@0.3.0 not found in yarn.lock but is-arrayish@0.3.2 was - looks like it's already updated (repository=renovate-reproductions/26670, branch=renovate/is-arrayish-0.x-lockfile)
DEBUG: Upgrade of is-arrayish to 0.3.2 is already done in existing branch (repository=renovate-reproductions/26670, branch=renovate/is-arrayish-0.x-lockfile)
DEBUG: No package files need updating (repository=renovate-reproductions/26670, branch=renovate/is-arrayish-0.x-lockfile)
DEBUG: Existing branch contains all necessary lock file updates (repository=renovate-reproductions/26670, branch=renovate/is-arrayish-0.x-lockfile)
DEBUG: No updated lock files in branch (repository=renovate-reproductions/26670, branch=renovate/is-arrayish-0.x-lockfile)
DEBUG: No files to commit (repository=renovate-reproductions/26670, branch=renovate/is-arrayish-0.x-lockfile)
```

**Reply by Garethp** [reply]
What I'm seeing when I run renovate with my config are these two commits:
 * https://github.com/Garethp/renovate-loop-bug/commit/8e994778ab796cb8e9f0de85e07c6cef25385d95
 * https://github.com/Garethp/renovate-loop-bug/commit/d94f177fd676c1d77eef479ce56785854c0f4de0

I notice your branch name doesn't match, and it looks like the branch is specific to just the one dependency. The error occurs when both dependencies are being updated in the same group/branch

**Reply by rarkins** [reply]
If it's different branch name then we're not reproducing the same thing. Maybe you have some config settings locally which affect it? Please look at https://github.com/renovate-reproductions/26670

**Reply by Garethp** [reply]
The repo looks the same and I don't have any env vars that should be affecting renovate. Are you applying the `config.js` from the repository when running it? That's where the branch/group name is being configured

**Reply by rarkins** [reply]
Please put all config in the renovate.json, and assume defaults in config.js

**Reply by Garethp** [reply]
Okay, it's in the `renovate.json` config and `config.js` are just defaults. Hopefully you'll be able to reproduce the bug easily now

**Reply by rarkins** [reply]
Yes, I can reproduce now. It seems it's the grouping which triggers it. 

Here's an example log from when the branch is fine/updated but Renovate checks and decides that it's not:

```
DEBUG: manager.getUpdatedPackageFiles() reuseExistingBranch=true (repository=renovate-reproductions/26670, branch=renovate/group-all)
DEBUG: npm.updateLockedDependency: is-arrayish@0.3.0 -> 0.3.2 [yarn.lock] (repository=renovate-reproductions/26670, branch=renovate/group-all)
DEBUG: is-arrayish@0.3.0 not found in yarn.lock but is-arrayish@0.3.2 was - looks like it's already updated (repository=renovate-reproductions/26670, branch=renovate/group-all)
DEBUG: Upgrade of is-arrayish to 0.3.2 is already done in existing branch (repository=renovate-reproductions/26670, branch=renovate/group-all)
DEBUG: npm.updateDependency(): dependencies.isarray = 2.0.5 (repository=renovate-reproductions/26670, branch=renovate/group-all)
DEBUG: No package files need updating (repository=renovate-reproductions/26670, branch=renovate/group-all)
DEBUG: Getting updated lock files (repository=renovate-reproductions/26670, branch=renovate/group-all)
```

As a workaround, see if you can run with `repositoryCache=enabled`. This avoids the cycling of the branch

**Reply by Garethp** [reply]
Unfortunately running with `repositoryCache=enabled` doesn't work, the cycling still happens

**Reply by rarkins** [reply]
Please give the debug log for the branch (i.e. any log message with branch=renovate/group-all)
